WIP:  Fix parallel test

nathanfranklin opened the pull request:
This PR looks to address #27

nathanfranklin commented on foam_controlwrapper/foam_runner.py:
```diff
@@ -27,24 +27,25 @@ def run(self):
-                  ' 2>&1'
-            subprocess.call(cmd, shell=True)
+                  os.path.join(self.case, 'log')
+            subprocess.check_call(cmd, shell=True)
```
@khiltunen , i have changed the use of ```subprocess.call``to ```subprocess.checkcall```.  now, we can see where things are failing:

```
[0] --> FOAM FATAL ERROR: 
[0] No coarse levels created, either matrix too small for GAMG or nCellsInCoarsestLevel too large.
    Either choose another solver of reduce nCellsInCoarsestLevel.
[0] 
[0]     From function GAMGSolver::GAMGSolver(const word& fieldName,const lduMatrix& matrix,const FieldField<Field, scalar>& interfaceBouCoeffs,const FieldField<Field, scalar>& interfaceIntCoeffs,const lduInterfaceFieldPtrsList& interfaces,const dictionary& solverControls)
[0]     in file matrices/lduMatrix/solvers/GAMG/GAMGSolver.C at line 121.
[0] 
FOAM parallel run exiting

```

Do you have any idea for what the problem is here?

nathanfranklin commented:
@khiltunen , have you had a chance to look at this?

nathanfranklin commented:
@khiltunen @jamartik ,   the issue #27 (which this PR was working on) is still open. For simphony-framework to use version 0.1.1 of this library, the testing of openfoam was disabled and an issue simphony/simphony-framework#35 was created.

nathanfranklin commented:
no, it occurs every time  (and is independent of symphony-framework).  See travis for this build for particular error message.

In general, the result of the subprocess calls (return code, error messages) were ignored so in commit   2da6892 , the results are now checked.   ````decomposePar```` was probably failing for some reason which resulted in the original error #27 (noticed on symphony-framework)and so this was the first step in checking that out.  but it turns out that `mpirun..` command is failing for the parallel tests.

nathanfranklin commented:
> no, it occurs every time (and is independent of symphony-framework). See travis for this build for particular error message.

@khiltunen , just to be clear, I mean this error:
````
[0] --> FOAM FATAL ERROR: 
[0] No coarse levels created, either matrix too small for GAMG or nCellsInCoarsestLevel too large.
    Either choose another solver of reduce nCellsInCoarsestLevel.
[0] 
[0]     From function GAMGSolver::GAMGSolver(const word& fieldName,const lduMatrix& matrix,const FieldField<Field, scalar>& interfaceBouCoeffs,const FieldField<Field, scalar>& interfaceIntCoeffs,const lduInterfaceFieldPtrsList& interfaces,const dictionary& solverControls)
[0]     in file matrices/lduMatrix/solvers/GAMG/GAMGSolver.C at line 121.
[0] 
FOAM parallel run exiting
````

khiltunen commented:
I tested this using command "python -m unittest discover" and get the same error message, but running it directly using python command "python foam_controlwrapper/tests/test_controlwrapper.py" tests are passed. This is strange. Do you have any clue what might lead to this?

nathanfranklin commented:
> I tested this using command "python -m unittest discover" and get the same error message, but running it directly using python command "python foam_controlwrapper/tests/test_controlwrapper.py" tests are passed. This is strange. Do you have any clue what might lead to this?

It fails for both commands for me (""python -m unittest discover" and ""python foam_controlwrapper/tests/test_controlwrapper.py"")

nathanfranklin commented:
@khiltunen , i just am realizing that i changed `foam_controlwrapper/tests/simplemesh.cuds` in commit e63229f.  That was a mistake (don't know how I did that).

nathanfranklin commented:
> @khiltunen , i just am realizing that i changed foam_controlwrapper/tests/simplemesh.cuds in commit e63229f. That was a mistake (don't know how I did that).

I have removed the commit that changed the `simplemesh.cuds` file.    This has no affect on the bug.

nathanfranklin commented:
Have you looked at the error message and then stepped through the debugger to see  what things look like after the ```decomposePar``` command.  I suspect there you will find out what is going on.

khiltunen commented:
@nathanfranklin i think i have managed to find where thinks go wrong. Firts of all the simplemesh.cuds file mesh has not enough levels to perform the decomposition. Also on that test example revealed a bug on foam templates for simpleFoam controlDict file. The writeControl was set to timeStep instead of runTime. This lead to situation that the last time step was not written and decomposePar might fail if any time step were written. So this bug fix needs to change two files (simplemesh.cuds and foam_templates.py). Is it better to do the changes directly to simphony-openfoam-0.1.3dev or how should we proceed ?

nathanfranklin commented:
Great work.

>  Is it better to do the changes directly to simphony-openfoam-0.1.3dev or how should we proceed ?

I don't totally understand what you meant by "change directly to simphony-openfoam-0.1.3".   We want to commit the additional changes you found to a branch, review them in a PR, and then merge the PR.  You can use this branch (as we need to include at least 2da6892 commit from this branch) or you can create a new branch.    

I just merged master into this branch.  Maybe that was your concern that this branch was created earlier (0.1.2dev).

nathanfranklin commented:
> Firts of all the simplemesh.cuds file mesh has not enough levels to perform the decomposition

@khiltunen @jmarcelogimenez 

I would sugget that the documentation include the limitations of using parallel processing (CUBAExt.NUMBER_OF_CORES] > 1).  Specifically:  
* in what situations (meshes that have certain characteristics etc..) paralle processing can be used with the FILE-IO wrapper
* and that the parallel processing is not available on the INTERNAL wrapper (https://github.com/simphony/simphony-openfoam/blob/590934b845730ffc5cd938b31d491078d312a115/openfoam-interface/internal-interface/wrapper/simphonyInterface.C#L1108)

I don't know if you want to address this with the bug fix or to do it at a later time.  Regardless, I would suggest opening an issue to keep track of it.

khiltunen commented:
Exactly that was my concern. I'll do the modifications to this branch.

khiltunen commented:
@nathanfranklin can you check that this is ready to merge

nathanfranklin commented:
:+1:
